037-pluggable-pod-security-profiles.md

```diff
@@ -0,0 +1,234 @@
+
+    default SecurityContext kafkaContainerSecurityContext(PodSecurityProviderContext context) {
+        return context.userSuppliedContext();
+    }
+
+    default SecurityContext kafkaInitContainerSecurityContext(PodSecurityProviderContext context) {
+        return context.userSuppliedContext();
+    }
+
+    // ...
+}
+```
+
+The interface will also have a `configure` method for configuring the provider.
+The `configure` method will get as a parameter an object describing the platform (Kubernetes version, features / APIs) which can be taken into account when creating the security context.
+This method will not have any default implementation.
+
+```java
+public interface PodSecurityProvider {
+    // ...
+    void configure(PlatformFeatures platformFeatures);
+    // ...
+}
+```
+
+`PlatformFeatures` in this method is another new interface which provides some of the methods provided by the `PlatformFeaturesAvailability` class which is part of the `operator-common` module`
+It uses the features which might be relevant for generating the security context:
+
+```java
+public interface PlatformFeatures {
+    KubernetesVersion getKubernetesVersion();
+    boolean isOpenshift();
+}
+```
+
+The `configure(...)` method can in addition to the `PlatformFeatures` also for example use environment variables to configure the provider.
+
+#### Implementations
+
+We will provide two implementations to support the baseline (default) and restricted profiles:
+* `BaselinePodSecurityProvider`
+* `RestrictedPodSecurityProvider`
+
+These implementations will be also part of the `api` module so that users can just extend them instead of implementing the provider from scratch.
+
+The `BaselinePodSecurityProvider` will implement the current behaviour:
+* Return the user-supplied security context when specified
+* When the user-supplied context is `null` and we are on OpenShift:
+    * A `null` will be returned (OpenShift injects its own Security Context in that case) for all Pods and containers
+* When the user-supplied context is `null` and we are not on OpenShift:
+    * Pods with storage will get a Security Context specifying the file system group as `0`
+    * All other pods will get `null` Security Context
+    * All containers will get `null` Security Context
+
+The `RestrictedPodSecurityProvider` will implement the following behaviour:
+* Return the user-supplied security context when specified
+* When the user-supplied context is `null`:
+    * Pods with storage will get a Security Context specifying the file system group as `0`
+    * All other pods will get `null` Security Context
+    * For the Kaniko container, an exception will be thrown since Kaniko cannot run under the Restricted profile
+    * All other containers will get a Security Context matching the _restricted_ Kubernetes profile:
+        * Disable privilage escalation
+        * Run as non-root
+        * Specify the Seccomp profile as `RuntimeDefault`
+        * Drop all capabilities
```

Comment:
Shouldn't we have a couple of examples here to show what is the security context configuration returned when using these two providers?

Reply:
I added the YAML examples of what they will return.